srml/babe/src/lib.rs

```diff
@@ -216,42 +262,53 @@ impl<T: Trait> OnTimestampSet<T::Moment> for Module<T> {
 	fn on_timestamp_set(_moment: T::Moment) { }
 }
 
-impl<T: Trait> session::OneSessionHandler<T::AccountId> for Module<T> {
-	type Key = AuthorityId;
+impl<T: Trait> session::ShouldEndSession<u64> for Module<T> {
+	fn should_end_session(_now: u64) -> bool {
+		LastSlotInEpoch::take()
+	}
+}
 
-	fn on_new_session<'a, I: 'a>(changed: bool, validators: I)
+impl<T: Trait + staking::Trait> session::OneSessionHandler<T::AccountId> for Module<T> {
+	type Key = AuthorityId;
+	fn on_new_session<'a, I: 'a>(_changed: bool, validators: I)
 		where I: Iterator<Item=(&'a T::AccountId, AuthorityId)>
 	{
-		// instant changes
-		if changed {
-			let next_authorities = validators.map(|(_, k)| k).collect::<Vec<_>>();
-			let last_authorities = <Module<T>>::authorities();
-			if next_authorities != last_authorities {
-				Self::change_authorities(next_authorities);
-			}
-		}
-
+		use staking::BalanceOf;
+		let to_votes = |b: BalanceOf<T>|
+			<T::CurrencyToVote as Convert<BalanceOf<T>, u64>>::convert(b);
 		let rho = UnderConstruction::get();
 		UnderConstruction::put([0; 32]);
 		let last_epoch_randomness = EpochRandomness::get();
 		let epoch_index = EpochIndex::get()
 			.checked_add(1)
 			.expect("epoch indices will never reach 2^64 before the death of the universe; qed");
 		EpochIndex::put(epoch_index);
-		EpochRandomness::put(NextEpochRandomness::get());
+		let next_authorities = validators.map(|(account, k)| {
+			(k, to_votes(staking::Module::<T>::stakers(account).total))
+		}).collect::<Vec<_>>();
+		let Epoch {
+			slot_number,
+			authorities,
+			randomness,
+		} = NextEpoch::get();
+		EpochRandomness::put(randomness);
+		Authorities::put(authorities);
 		let mut s = [0; 72];
 		s[..32].copy_from_slice(&last_epoch_randomness);
 		s[32..40].copy_from_slice(&epoch_index.to_le_bytes());
 		s[40..].copy_from_slice(&rho);
-		NextEpochRandomness::put(runtime_io::blake2_256(&s))
+		let randomness = runtime_io::blake2_256(&s);
+		Self::change_authorities(Epoch {
+			randomness,
+			authorities: next_authorities,
+			slot_number: slot_number
```

Comment:
And here you calculate it in a different way?

Reply:
Fixed by deleting the `slot_number` field.